dd-sdk-android-core/src/main/kotlin/com/datadog/android/core/internal/persistence/tlvformat/FileTLVBlockReader.kt

```diff
@@ -0,0 +1,126 @@
+/*
+ * Unless explicitly stated otherwise all files in this repository are licensed under the Apache License Version 2.0.
+ * This product includes software developed at Datadog (https://www.datadoghq.com/).
+ * Copyright 2016-Present Datadog, Inc.
+ */
+
+package com.datadog.android.core.internal.persistence.tlvformat
+
+import androidx.annotation.WorkerThread
+import com.datadog.android.api.InternalLogger
+import com.datadog.android.core.internal.persistence.datastore.ext.toInt
+import com.datadog.android.core.internal.persistence.datastore.ext.toShort
+import com.datadog.android.core.internal.persistence.file.FileReaderWriter
+import com.datadog.android.core.internal.persistence.file.existsSafe
+import com.datadog.android.core.internal.persistence.file.lengthSafe
+import java.io.File
+import java.io.IOException
+import java.io.InputStream
+import java.util.Locale
+import java.util.concurrent.atomic.AtomicBoolean
+
+internal class FileTLVBlockReader(
+    val internalLogger: InternalLogger,
+    val fileReaderWriter: FileReaderWriter
+) {
+    private var endOfStream = AtomicBoolean(false)
+
+    @WorkerThread
+    internal fun all(file: File): List<TLVBlock> {
+        if (!file.existsSafe(internalLogger) || file.lengthSafe(internalLogger) == 0L) {
+            return arrayListOf()
+        }
+
+        val blocks = mutableListOf<TLVBlock>()
+        val stream = fileReaderWriter.readData(file).inputStream()
+
+        while (!endOfStream.get()) {
+            val nextBlock = readBlock(stream) ?: break
+            blocks.add(nextBlock)
+        }
+
+        return blocks
+    }
+
+    private fun readBlock(stream: InputStream): TLVBlock? {
+        val type = readType(stream) ?: return null
+
+        val data = readData(stream)
+
+        return TLVBlock(type, data)
+    }
+
+    private fun readType(stream: InputStream): TLVBlockType? {
+        val typeBlockSize = UShort.SIZE_BYTES
+        val bytes = read(stream, typeBlockSize)
+
+        if (bytes.size != typeBlockSize) {
+            return null
+        }
+
+        val shortValue = bytes.toShort()
+        val tlvHeader = TLVBlockType.fromValue(shortValue.toUShort())
+
+        if (tlvHeader == null) {
+            logTypeCorruptionError(shortValue)
+        }
+
+        return tlvHeader
+    }
+
+    private fun readData(stream: InputStream): ByteArray {
```

Comment:
if we read everything in one buffer isn't it going to blow up the RAM memory ? I am only asking because usually with buffers we should work with smaller sizes in order to maintain the performance. Not sure if we can avoid having all the data in memory at once here though.

Reply:
I've changed the inputStream to be buffered. It's possible that in an image heavy application we could have a very large number of resource hashes being written every time, and we're in any case keeping them always in memory in order to avoid sending them to the backend. If we see that this causes a performance hit we may need to rethink this mechanism (also on ios). If I'm not mistaken every md5 hash is 16 bytes, so we would have around 65000 Strings per mb, and we can try to optimize the encoding to use less space.

Reply:
On ios I see that they've put in a maximum length of 10mb per entry and if it attempts to exceed this size then it throws an exception - I'll implement a similar mechanism to prevent extremely large edge cases.

Reply:
Actually you don't need to have an input stream at all. Input stream is designed to back potentially infinite data with a cost of handling a bunch of possible `IOException` types.

Here you have everything as a `ByteArray` directly if I'm not wrong, so everything is already loaded in the memory and it finite, you can work with an array directly, without coming to the streams.